OPAL/ide/src/main/scala/org/opalj/ide/solver/IDEAnalysis.scala

```diff
@@ -0,0 +1,1115 @@
+            nodeWorkList.isEmpty
+        }
+
+        def getNodeWorkListSize: Int = {
+            nodeWorkList.size
+        }
+
+        def getValue(node: Node, callable: Callable): Value = {
+            values.getOrElse(callable, mutable.Map.empty).getOrElse(node, problem.lattice.top) // else part handles IDE line 1
+        }
+
+        def getValue(node: Node): Value = {
+            getValue(node, icfg.getCallable(node._1))
+        }
+
+        def setValue(node: Node, newValue: Value, callable: Callable): Unit = {
+            values.getOrElseUpdate(callable, { mutable.Map.empty }).put(node, newValue)
+        }
+
+        def setValue(node: Node, newValue: Value): Unit = {
+            setValue(node, newValue, icfg.getCallable(node._1))
+        }
+
+        def clearValues(): Unit = {
+            values.clear()
+        }
+
+        /**
+         * @return a map from statements to results and the results for the callable in total (both per requested
+         *         callable)
+         */
+        def collectResults(callables: collection.Set[Callable]): Map[
+            Callable,
+            (collection.Map[Statement, collection.Set[(Fact, Value)]], collection.Set[(Fact, Value)])
+        ] = {
+            val valuesByCallable = callables.map { callable =>
+                callable -> values.getOrElse(callable, immutable.Map.empty[Node, Value])
+            }.toMap
+
+            valuesByCallable.map { case (callable, values) =>
+                val resultsByStatement = values
+                    .view
+                    .filterKeys { case (n, d) => d != problem.nullFact }
+                    .groupMap(_._1._1) { case ((_, d), value) => (d, value) }
+                    .map { case (n, dValuePairs) =>
+                        (
+                            n,
+                            dValuePairs.groupMapReduce(_._1)(_._2) { (value1, value2) =>
+                                problem.lattice.meet(value1, value2)
+                            }.toSet
+                        )
+                    }
+
+                val resultsForExit = resultsByStatement
+                    .filter { case (n, _) => icfg.isNormalExitStatement(n) }
+                    .map(_._2.toList)
+                    .flatten
+                    .groupMapReduce(_._1)(_._2) {
+                        (value1, value2) => problem.lattice.meet(value1, value2)
+                    }
+                    .toSet
+
+                callable -> (resultsByStatement, resultsForExit)
+            }
+        }
+
+        def addDependee(eOptionP: SomeEOptionP, c: () => Unit): Unit = {
+            // The eOptionP is only inserted the first time the corresponding EPK occurs. Consequently, it is the most
+            // precise property result that is seen by all dependents.
+            val (_, set) = dependees.getOrElseUpdate(eOptionP.toEPK, (eOptionP, mutable.Set.empty))
+            set.add(c)
+        }
+
+        def areDependeesEmpty: Boolean = {
```

Comment:
It would be more conventional to have a `hasDependees` method with inverted semantics